### PR TITLE
Add JWT auth middleware

### PR DESCRIPTION
**Pull request by welteki**
## Description

Add JWT authentication middleware.

The JWT middleware can be used to authorize http request for functions with IAM for OpenFaaS. The middleware verifies the permissions in the `function` claim of an OpenFaaS function access token that is set in the `Authorization` header of the request.

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
- [ ] My issue has received approval from the maintainers or lead with the `design/approved` label

Move over the JWT authentication middleware from the of-watchdog so that it can easily be shared between the different watchdog implementations.

This change is part of adding [built-in authentication for functions](https://www.openfaas.com/blog/built-in-function-authentication/) with IAM for OpenFaaS.

## How Has This Been Tested?

The JWT authentication middleware was originally part of the of-watchdog and has been tested there E2E.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
I`OPNEFAAS_NAME` is a typo in the PR description. Does it appear anywhere else?

I would not expect any middleware components to be configured via environment variables, but by proper struct constructors or methods.

File: function-auth/jwt_authenticator.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
+package functionauth
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: What if this was just auth?

File: function-auth/jwt_authenticator.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+		return nil, err
+	}
+
+	if opts.Debug {
+		for _, key := range keySet.Keys {
+			log.Printf("[JWT Auth] Key: %s", key.KeyID)
+		}
+	}
+
+	fn := func(next http.Handler) http.Handler {
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: Inline this?

File: function-auth/jwt_authenticator.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/rakutentech/jwk-go/jwk"
+)
+
+const (
+	authorityURL      = "http://gateway.openfaas:8080/.well-known/openid-configuration"
+	localAuthorityURL = "http://127.0.0.1:8000/.well-known/openid-configuration"
+	functionRealm     = "IAM function invoke"
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: I'm not sure on the way the wording reads.

Comment (alexellis) [reply function-auth/jwt_authenticator.go]: I saw [an example like](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) `WWW-Authenticate: Basic realm="Access to the staging site", charset="UTF-8"`

File: function-auth/jwt_authenticator.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+			if k.KeyID == kid {
+				key = &k
+				break
+			}
+		}
+
+		if key == nil {
+			return nil, fmt.Errorf("invalid kid: %s", kid)
+		}
+		return key.Key.(crypto.PublicKey), nil
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: Do you need to run a check before accessing as this type?

`if ok, k:= .... {`

Comment (alexellis) [reply function-auth/jwt_authenticator.go]:
```
		var signingKey crypto.PublicKey
		if v, ok := c.keySpec.Key.(crypto.PublicKey); ok {
			signingKey = v
		} else {
			return OpenFaaSClaims{}, fmt.Errorf("unable to convert key to crypto.PublicKey")
		}
```

From faas-netes

File: function-auth/jwt_authenticator.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+		log.Printf("%s: %v", issuer, key.KeyID)
+	}
+
+	var bearer string
+	if v := r.Header.Get("Authorization"); v != "" {
+		bearer = strings.TrimPrefix(v, "Bearer ")
+	}
+
+	if bearer == "" {
+		httpUnauthorized(w, "Bearer must be present in Authorization header")
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: writeUnauthorized

writeUnauthorizedRealm

File: function-auth/jwt_authenticator.go [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+		if a.opts.Debug {
+			log.Printf("[JWT Auth] Token: audience: %v\tissuer: %v", functionClaims.Audience, functionClaims.Issuer)
+		}
+
+		kid, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, fmt.Errorf("invalid kid: %v", token.Header["kid"])
+		}
+		var key *jwk.KeySpec
+		for _, k := range a.keySet.Keys {
```
Comment (alexellis) [inline function-auth/jwt_authenticator.go]: Do keys refresh? This may be needed for when the gateway gets restarted.

Comment (alexellis) [reply function-auth/jwt_authenticator.go]: Having discussed this, OK to leave for future. Leave a TODO note:

`HV: ...`

File: function-auth/jwt_authenticator.go (hunk 4 above)
Comment (welteki) [reply function-auth/jwt_authenticator.go]: After looking into this, there is actually no need to cast and check the key type here. The jwt package already returns clear errors when the KeyFunc returns an interface of the wrong type.

File: function-auth/jwt_authenticator.go (hunk 5 above)
Comment (welteki) [reply function-auth/jwt_authenticator.go]: The function has been renamed to `writeUnauthorized`.

File: function-auth/jwt_authenticator.go (hunk 1 above)
Comment (welteki) [reply function-auth/jwt_authenticator.go]: The package has been renamed to auth.

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**